### PR TITLE
Fix ICE caused by Drop implementations for unsized types

### LINKED ISSUE #26709: ICE with type-pruned complex smart pointer
Converting my `Aref` type to handle unsized types has exposed an ICE (assertion failure) in `librustc_trans/trans/glue.rs` when handling `ArefInner`'s  destructor.

```
error: internal compiler error: unexpected panic
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'assertion failed: `(left == right)` (left: `2`, right: `1`)', ../src/librustc_trans/trans/glue.rs:358


playpen: application terminated with error code 101
```

Reproduction code (triggers assertion on playpen, and current nightly `rustc 1.3.0-nightly (faa04a8b9 2015-06-30)`)
```rust
#![feature(unsize,coerce_unsized,unsafe_no_drop_flag)]
use std::{ops,marker};
use std::sync::atomic::{Ordering,AtomicUsize};

pub struct Aref<T: ?Sized> {
    __inner: Box<ArefInner<T>>,
}
// Allow coercing
impl<T: ?Sized + marker::Unsize<U>, U: ?Sized> ops::CoerceUnsized<Aref<U>> for Aref<T> {}

#[unsafe_no_drop_flag]
pub struct ArefInner<T: ?Sized> {
        count: AtomicUsize,
        data: T,
}

impl<T: ?Sized> ops::Drop for ArefInner<T> {
        fn drop(&mut self) {
                assert_eq!(self.count.load(Ordering::Relaxed), 0);
        }
}

impl<T> Aref<T> {
        pub fn new(val: T) -> Aref<T> {
                Aref {
                        __inner: Box::new(unsafe{ ArefInner::new(val) }),
                }
        }
}

impl<T> ArefInner<T> {
        pub unsafe fn new(val: T) -> ArefInner<T> {
                ArefInner {
                        count: AtomicUsize::new(0),
                        data: val,
                }
        }
}


fn main() {
    let ar = Aref::new(1234);
    let ar_disp: Aref<::std::fmt::Display> = ar;
}

```

### PR DESCRIPTION
**Pull request by dotdash**
Fixes #26709

### CODE REVIEW

File: src/test/run-pass/issue26709.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+struct Wrapper<T: ?Sized>(T);
+
+impl<T: ?Sized> Drop for Wrapper<T> {
+    fn drop(&mut self) {}
```
Comment (arielb1) [inline src/test/run-pass/issue26709.rs]: Check that the destructor is really called correctly.

Comment (pnkfelix) [inline src/test/run-pass/issue26709.rs]: @dotdash yes, @arielb1 is right; in simple programs like this, it suffices to add a global counter and make sure it has the right values after each expected drop.

**Comment by pnkfelix** [comment]
r=me once the test is expanded as described above.

File: src/test/run-pass/issue26709.rs (hunk 1 above)
Comment (dotdash) [inline src/test/run-pass/issue26709.rs]: I went for a check that uses a pointer stored in the struct, to make sure that `drop()` is called with the right arguments.

**Pull request merged**